lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h

```diff
@@ -219,6 +219,187 @@ class ConvertTritonGPUOpToLLVMPatternBase {
+                        Type resElemTy, SharedMemoryObject smemObj,
+                        ConversionPatternRewriter &rewriter,
+                        SmallVectorImpl<Value> &offsetVals,
+                        SmallVectorImpl<Value> &srcStrides) const {
+    // This utililty computes the pointers for accessing the provided swizzled
+    // shared memory layout `resSharedLayout`. More specifically, it computes,
+    // for all indices (row, col) of `srcEncoding` such that idx % inVec = 0,
+    // the pointer: ptr[(row, col)] = base + (rowOff * strides[ord[1]] + colOff)
+    // where :
+    //   compute phase = (row // perPhase) % maxPhase
+    //   rowOff = row
+    //   colOff = colOffSwizzled + colOffOrdered
+    //     colOffSwizzled = ((col // outVec) ^ phase) * outVec
+    //     colOffOrdered = (col % outVec) // minVec * minVec
+    //
+    // Note 1:
+    // -------
+    // Because swizzling happens at a granularity of outVec, we need to
+    // decompose the offset into a swizzled factor and a non-swizzled (ordered)
+    // factor
+    //
+    // Note 2:
+    // -------
+    // If we have x, y, z of the form:
+    // x = 0b00000xxxx
+    // y = 0byyyyy0000
+    // z = 0b00000zzzz
+    // then (x + y) XOR z = 0byyyyxxxx XOR 0b00000zzzz = (x XOR z) + y
+    // This means that we can use some immediate offsets for shared memory
+    // operations. More specifically
+    auto dstPtrTy = ptr_ty(resElemTy, 3);
+    auto dstOffset = dot(rewriter, loc, offsetVals, smemObj.strides);
+    Value dstPtrBase = gep(dstPtrTy, smemObj.base, dstOffset);
+
+    auto srcEncoding = srcTy.getEncoding();
+    auto srcShape = srcTy.getShape();
+    unsigned numElems = triton::gpu::getElemsPerThread(srcTy);
+    // swizzling params as described in TritonGPUAttrDefs.td
+    unsigned outVec = resSharedLayout.getVec();
+    unsigned perPhase = resSharedLayout.getPerPhase();
+    unsigned maxPhase = resSharedLayout.getMaxPhase();
+    // order
+    auto inOrder = triton::gpu::getOrder(srcEncoding);
+    auto outOrder = triton::gpu::getOrder(resSharedLayout);
+    // tensor indices held by the current thread, as LLVM values
+    auto srcIndices = emitIndices(loc, rewriter, srcEncoding, srcShape);
+    // return values
+    DenseMap<unsigned, Value> ret;
+    // cache for non-immediate offsets
+    DenseMap<unsigned, Value> cacheCol, cacheRow;
+    unsigned minVec = std::min(outVec, inVec);
+    for (unsigned elemIdx = 0; elemIdx < numElems; elemIdx += minVec) {
+      // extract multi dimensional index for current element
+      auto idx = srcIndices[elemIdx];
+      Value idxCol = idx[inOrder[0]]; // contiguous dimension
+      Value idxRow = idx[inOrder[1]]; // discontiguous dimension
+      Value strideCol = srcStrides[inOrder[0]];
+      Value strideRow = srcStrides[inOrder[1]];
+      // extract dynamic/static offset for immediate offseting
+      unsigned immedateOffCol = 0;
+      if (auto add = dyn_cast_or_null<LLVM::AddOp>(idxCol.getDefiningOp()))
```

Comment:
What if `add` if NULL?

Reply:
Maybe add `assert` to give warnings?

Reply:
then `if (auto add = dyn_cast_or_null<LLVM::AddOp>(idxCol.getDefiningOp()))` will evaluate to `if(0)` and the branch won't be taken right?

Reply:
Yeah, I meant then `immedateOffCol` is will be just 0, is it correct understand all circumstances? Even when `_cst` is a dynamic value?

Reply:
I think so. `Value idxCol = idx[inOrder[0]];` is the default value that should always work. Then, if it has the form `idxCol = colDyn + colCst` then we extract the constant and override `idxCol` with `colDyn`. Do you have in mind a case where it wouldn't work?

Reply:
Never mind, just checked the code, it probably should always work since we use 
`multiDimIdx[n][k] = add(multiDimBase[k], idx_val(offset[n][k]));`